osu.Server.Spectator/Hubs/Metadata/MetadataHub.cs

```diff
@@ -107,13 +113,59 @@ public async Task UpdateStatus(UserStatus? status)
             }
         }
 
+        private static readonly object update_stats_lock = new object();
+
         public async Task<MultiplayerPlaylistItemStats[]> BeginWatchingMultiplayerRoom(long id)
         {
             await Groups.AddToGroupAsync(Context.ConnectionId, MultiplayerRoomWatchersGroup(id));
             await scoreProcessedSubscriber.RegisterForMultiplayerRoomAsync(Context.GetUserId(), id);
 
             using var db = databaseFactory.GetInstance();
-            return await db.GetMultiplayerRoomStatsAsync(id);
+
+            MultiplayerRoomStats stats = (await cache.GetOrCreateAsync<MultiplayerRoomStats>(id.ToString(), e =>
+            {
+                e.SlidingExpiration = TimeSpan.FromMinutes(30);
+                return Task.FromResult(new MultiplayerRoomStats { RoomID = id });
+            }))!;
+
+            await updateMultiplayerRoomStatsAsync(db, stats);
+
+            // Outside of locking so may be mid-update, but that's fine we don't need perfectly accurate for client-side.
+            return stats.PlaylistItemStats.Values.ToArray();
+        }
+
+        private async Task updateMultiplayerRoomStatsAsync(IDatabaseAccess db, MultiplayerRoomStats stats)
+        {
+            long[] playlistItemIds = (await db.GetAllPlaylistItemsAsync(stats.RoomID)).Select(item => item.id).ToArray();
+
+            for (int i = 0; i < playlistItemIds.Length; ++i)
+            {
+                long itemId = playlistItemIds[i];
+
+                if (!stats.PlaylistItemStats.TryGetValue(itemId, out var itemStats))
+                    stats.PlaylistItemStats[itemId] = itemStats = new MultiplayerPlaylistItemStats { PlaylistItemID = itemId, };
+
+                SoloScore[] scores = (await db.GetScoresForPlaylistItem(itemId, itemStats.LastProcessedScoreID)).ToArray();
+
+                if (scores.Length == 0)
+                    return;
+
+                // Lock globally for simplicity.
+                // If it ever becomes an issue we can move to per-item locking or something more complex.
+                lock (update_stats_lock)
```

Comment:
I don't think this lock does much to prevent double-counting without a control read of `LastProcessedScoreID` inside.

Scenario I'm thinking of:

```
thread A					thread B

db.GetScoresForPlaylistItem(itemId, 1000)

						db.GetScoresForPlaylistItem(itemId, 1000)

received scores from id 1000 to 1005
enters update_stats_lock
does incremental updates for above scores
updates LastProcessedScoreID = 1005
exits update_stats_lock

						received scores from id 1000 to 1005
						enters update_stats_lock
						does a second incremental update for the same scores
						updates LastProcessedScoreID = 1005
						exits lock
```

A check that `LastProcessedScoreID` inside the critical section would presumably prevent this:

```
thread A					thread B

db.GetScoresForPlaylistItem(itemId, 1000)

						db.GetScoresForPlaylistItem(itemId, 1000)

received scores from id 1000 to 1005
enters update_stats_lock
checks LastProcessedScoreID == 1000 (pass)
does incremental updates for above scores
updates LastProcessedScoreID = 1005
exits update_stats_lock

						received scores from id 1000 to 1005
						enters update_stats_lock
						checks LastProcessedScoreID == 1000 (fail)
						exits lock
```